python/nano/src/bigdl/nano/pytorch/lightning.py

```diff
@@ -0,0 +1,66 @@
+#
+# Copyright 2016 The BigDL Authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+from pytorch_lightning import LightningModule
+from torch import nn
+import torch
+from torch.nn.modules.loss import _Loss
+from torch.optim import Optimizer
+
+
+class LightningModuleFromTorch(LightningModule):
+    def __init__(self, model: nn.Module, loss: _Loss, optimizer: Optimizer):
+        """
+        Integrate pytorch modules, loss, optimizer to pytorch-lightning model.
+
+        :param model:       Pytorch model to be converted.
+        :param loss:        A torch loss function.
+        :param optimizer:   A torch optimizer.
+        """
+        super().__init__()
+        self._copy(model)
+        self.loss = loss
+        self.optimizer = optimizer
+
+    def _copy(self, torch_model: nn.Module):
```

Comment:
Is it safe to do this? 
Looks a little hacking to me. 
Should we stick to the public API pytorch-lightning provides?

Reply:
Why not make `torch_model` a member variable?

Reply:
We have `load_state_dict` in `nn.Module`. So I'm concerned if we make it as a member variable user won't be able to load a pretrained torch model to pytorch-lightning. We can still assume that user has already loaded the parameters before creating this pl model. But I'm not sure if all other functions work fine.

Reply:
Having checked Pytorch API doc, it seems only `load_state_dict` is influenced since the module dict will be changed if make it a member variable. I think we have 3 options:

1.  Make pytorch model as a member variable and assume user has already loaded the model.
2.  Make pytorch model as a member variable and override `load_state_dict`.
3.  Move `_copy` into `__init__`. (Not sure if this seems not that hacking.)

Reply:
We can do #2.

Another option: can we use `__getattr__`, similar to https://github.com/intel-analytics/BigDL/blob/branch-2.0/python/friesian/src/bigdl/friesian/feature/table.py#L663?

Reply:
No idea how `getattr` works in this case. Current implementation will try to load self.model firstly and then try to load self if failed.

```python
try:
    self.model.load_state_dict(state_dict)
except RuntimeError:
    super().load_state_dict(state_dict)
```
And I added two unit tests to make sure they work in both cases.

Reply:
> No idea how `getattr` works in this case. Current implementation will try to load self.model firstly and then try to load self if failed.
> 
> ```python
> try:
>     self.model.load_state_dict(state_dict)
> except RuntimeError:
>     super().load_state_dict(state_dict)
> ```
> 
> And I added two unit tests to make sure they work in both cases.

Exactly what `Table` does:
``` 
   def __getattr__(self, name):
        return self.model.__getattr__(name)
```

Reply:
I understand now. If we use this method, it works when I load state_dict from a torch model checkpoint. 
Let's say keys are "conv.weight", "conv.bias", and keys for lightning module are "model.conv.weight", "model.conv.bias", it works.
Otherwise if I stored a pytorch-lightning model, its keys are actually "model.conv.weight", "model.conv.bias", then it failed. 
If we just want to resolve `load_state_dict`, possibly:

1.  We accept that user can load state dict from either a torch model or pl model.
Simplest way it to try loading into `self.model` and then `self`, otherwise raise error.

2.  We assume user always load state dict from torch model before calling `compile` or `LightningModuleFromTorch`.
Then we don't have to do anything.

Reply:
> I understand now. If we use this method, it works when I load state_dict from a torch model checkpoint. Let's say keys are `conv.weight`, `conv.bias`, and keys for lightning module are `model.conv.weight`, `model.conv.bias`, it works. Otherwise if I stored a pytorch-lightning model, its keys are actually `model.conv.weight`, `model.conv.bias`, then it failed. If we just want to resolve `load_state_dict`, possibly:
> 
> 1. We accept that user can load state dict from either a torch model or pl model.
>    Simplest way it to try loading into `self.model` and then `self`, otherwise raise error.
> 2. We assume user always load state dict from torch model before calling `compile` or `LightningModuleFromTorch`.
>    Then we don't have to do anything.

Main difference is keys for a pl model have 'model.' prefix.

Reply:
> I understand now. If we use this method, it works when I load state_dict from a torch model checkpoint. Let's say keys are `conv.weight`, `conv.bias`, and keys for lightning module are `model.conv.weight`, `model.conv.bias`, it works. Otherwise if I stored a pytorch-lightning model, its keys are actually `model.conv.weight`, `model.conv.bias`, then it failed. If we just want to resolve `load_state_dict`, possibly:
> 
> 1. We accept that user can load state dict from either a torch model or pl model.
>    Simplest way it to try loading into `self.model` and then `self`, otherwise raise error.
> 2. We assume user always load state dict from torch model before calling `compile` or `LightningModuleFromTorch`.
>    Then we don't have to do anything.

I think the point is:
1) We can override whatever methods (such as `load_state_dict`, `forward`, etc.) that we would like to in `LightningModuleFromTorch`
2) For methods defined in the original python module that we do not override, we may use `__getattr__` so that these methods can still work (if needed at all)

Reply:
Got it. We discussed and preferred 2nd option.
Everything looks good now.